lift/Readme.md

```diff
@@ -6,271 +6,360 @@
-├── datamodel.mdl
-└── prisma.yml
+migrations/
+├── 001_setup.down.sql
+├── 001_setup.up.sql
+├── 002_convo.down.sql
+├── 002_convo.up.sql
+├── 003_tempo.down.sql
+├── 003_tempo.up.sql
+├── 004_eventids.down.sql
+├── 004_eventids.up.sql
+├── 005_update_pricing.down.sql
+├── 005_update_pricing.up.sql
 ```
 
-Executing `prisma migrate` will create the following folder structure and migrate the underlying database structure:
+The numbers determine the order the migrations should be performed in. We first run `001`, then `002`, etc. The `up` and `down` determine the direction we're
+migrating. If we're migrating up, we'll run the `up` scripts, if we're migrating down, we'll run the `down` scripts.
+
+Traditionally, you write migrations by hand to migrate your database.
 
+```sql
+alter type report_status rename to report_status_old;
+create type report_status as enum('ASKED','SKIP','COMPLETE','PENDING');
+alter table reports alter column "status" drop default;
+alter table reports alter column "status" type report_status using "status"::text::report_status;
+alter table reports alter column "status" set default 'ASKED';
+drop type report_status_old;
 ```
-.
-├── datamodel.mdl
-├── migrate
-│   └── 20190322092247
-│       └── datamodel.mdl
-└── prisma.yml
+
+This is error-prone and stressful, especially when you're operating on your production data.
+
+## The Lift Approach
+
+Lift works differently. While Lift still has a `migrations/` folder, the migrations are generated for you.
+
+```diff
+model Blog {
+  id         Int      @id
++  website    String   @unique
+  posts      Post[]
+  created_at DateTime
+}
+
+model Post {
+  id         Int       @id
+-  title      String
++  slug      String
+  author_id  User?
+  blog_id    Blog?
+  comments   Comment[]
+  created_at DateTime
+}
+
++model Comment {
++  id         Int      @id
++  post_id    Post?
++  comment    String
++  created_at DateTime
++}
 ```
 
-So in addition to migrating the database, `prisma migrate` now also creates a new migration folder with the timestamp of the migration.
+With Lift, you just need to change your `schema.prisma` file and run `lift save`. This will generate the necessary steps to transition your schema from A to B.
 
-## Motivation
+## Concepts
 
-These are the main reasons we want to add migration files:
+Lift has the following concepts: projects, migrations, steps, and hooks.
 
-- Being able to run arbitrary data migration scripts
-- Having SQL / other database native escape hatches to perform structural database changes which Prisma doesn't support yet
+- A project has many migrations
+- A migration has many steps
+- A migration has many hooks
 
-## Terminology
+### Step
 
-Assuming that the reader knows, what Prisma 1.x looks like, the following terms are new concepts which we introduce with this spec.
+A Step is a single command that we'll run against the database. Steps describe models, fields and relationships generically, so they can be easily translated to
```

Comment:
This sounds like as if one step would always lead to one command against the database. But this is not the case. Sometimes a step does not even trigger a single database command, e.g. `CreateEnum` which does not have a representation in the database at the moment. Often multiple steps are grouped into one command, e.g. a `CreateModel` and multiple `CreateField` result in one `CREATE TABLE` in the sql connector.